plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/CabalAdd.hs

```diff
@@ -0,0 +1,160 @@
+        title = "Add dependency " <> suggestedDep <> " " <> versionTitle
+
+        version = if T.null suggestedVersion then Nothing else Just suggestedVersion
+        params = CabalAddCommandParams {cabalPath = cabalFile, dependency = suggestedDep, version=version}
+        command = mkLspCommand plId (CommandId cabalAddCommand) "Execute Code Action" (Just [toJSON params])
+      in CodeAction title (Just CodeActionKind_QuickFix) (Just []) Nothing Nothing Nothing (Just command) Nothing
+
+-- | Gives a mentioned number of hidden packages given
+--   a specific error message
+missingDependenciesSuggestion :: Int -> T.Text -> [(T.Text, T.Text)]
+missingDependenciesSuggestion maxCompletions msg = take maxCompletions $ getMatch (msg =~ regex)
+  where
+    regex :: T.Text -- TODO: Support multiple packages suggestion
+    regex = "Could not load module \8216.*\8217.\nIt is a member of the hidden package \8216([a-z]+)[-]?([0-9\\.]*)\8217"
+    getMatch :: (T.Text, T.Text, T.Text, [T.Text]) -> [(T.Text, T.Text)]
+    getMatch (_, _, _, []) = []
+    getMatch (_, _, _, [dependency]) = [(dependency, T.empty)]
+    getMatch (_, _, _, [dependency, version]) = [(dependency, version)]
+    getMatch (_, _, _, _) = error "Impossible pattern matching case"
+
+
+hiddenPackageAction :: Int -> Uri -> Diagnostic -> [CodeAction]
+hiddenPackageAction = undefined
+
+hiddenPackageSuggestion :: Int -> T.Text -> [T.Text]
+hiddenPackageSuggestion maxCompletions msg = undefined
+
+cabalAddCommand :: IsString p => p
+cabalAddCommand = "cabalAdd"
+
+data CabalAddCommandParams =
+     CabalAddCommandParams { cabalPath  :: FilePath
+                           , dependency :: T.Text
+                           , version    :: Maybe T.Text
+                           }
+  deriving (Generic, Show)
+  deriving anyclass (FromJSON, ToJSON)
+
+command :: CommandFunction IdeState CabalAddCommandParams
+command _ _ (CabalAddCommandParams {cabalPath = path, dependency = dep, version = mbVer}) = do
+  let specifiedDep = case mbVer of
+        Nothing  -> dep
+        Just ver -> dep <> " ^>=" <> ver
+  void $ liftIO $ addDependency path (fromList [T.unpack specifiedDep])
+  pure $ InR Null
+
+-- | Gives cabal file's contents or throws error.
+--   Inspired by @readCabalFile@ in cabal-add,
+--   Distribution.Client.Main
+readCabalFile :: FilePath -> IO ByteString
+readCabalFile fileName = do
+  cabalFileExists <- doesFileExist fileName
+  if cabalFileExists
+    then snd . patchQuirks <$> B.readFile fileName
+    else error ("Failed to read cabal file at " <> fileName)
+
+-- | Constructs prerequisets for the @executeConfig@
+--   and runs it, given path to the cabal file and
+--   a dependency message.
+--
+--   Inspired by @main@ in cabal-add,
+--   Distribution.Client.Main
+addDependency :: FilePath -> NonEmpty String -> IO ()
+addDependency cabalFilePath dependency = do
+  let rcnfComponent = Nothing
```

Comment:
This is a good first step, but it would be nice to cover components other than the main library. Is there a way to ask Cabal which component a certain file belongs to?

Reply:
Personally, I don't know any convenient tools like this.
I think it can be done by parsing the cabal file and looking, where its module name is mentioned. But this approach will not be trivial to implement.

Reply:
Could https://hackage.haskell.org/package/Cabal-3.12.1.0/docs/Distribution-Simple-BuildTarget.html#t:BuildTarget be of any help? Presumably it should be possibly to pass file name to `readBuildTargets` (?) and get `BuildTargetModule ComponentName ModuleName`, where `ComponentName` is what we are looking for.

Reply:
This is a great suggestion and seems to be working on small projects at least!

Reply:
I am working on the integration of this function, but I faced some problems.
Unfortunately, `readBuildTargets` does not resolve them completely and makes it impossible to use the found target.

For example, if there is a library and an executable with the same name, it will not resolve them into `["exe:common-name", "lib:common-name"]` and will just give `["common-name"]`. (oversimplified demonstration).

It will be impossible to use just "common-name" since it's not resolved.

Is there a way to just get all the targets? Or a way to resolve them? That would be very convenient.
I think it's possible to bruteforce them, but I would rather avoid it.

Reply:
@fendor @michaelpj is there any ready-to-use machinery for this in HLS? I'd imagine determining component by filename is a common task.

Catching `VerboseException` is not too bad, I think.

Reply:
>  I'd imagine determining component by filename is a common task.

No, as HLS is build-tool independent, there are very few locations where we handle build-tools differently and the hls-cabal-plugin is young by comparison to the other infrastructure.

Reply:
I discussed with @VenInf a possible solution